### PR TITLE
231 refactoring

### PR DESCRIPTION
**Pull request by fishermans**
Hi @dubreuia,

I spent some more time keeping this wonderful plugin updated.

Main change is moving to services and removing component implementation regarding [https://plugins.jetbrains.com/docs/intellij/dynamic-plugins.html](https://plugins.jetbrains.com/docs/intellij/dynamic-plugins.html).

I also implemented a FileDocumentManagerListener and registering it to the EP which should be a cleaner way then using the message bus. At this point, we can decide whether to register multiple Listener for each project or using an application based listener for all projects and iterating over the projects. As long as changed documents are held systemwide, it would not make any difference from my-point-of-view.

I tested it on Idea, RubyMine and Resharper.

Please review my changes when you find some time.

BR
Tiemo

### CODE REVIEW

**Comment by fishermans** [comment]
Should also fix #384

File: src/main/java/com/dubreuia/core/component/Engine.java [hunk 1]
```diff
@@ -77,9 +81,12 @@
...
     }
 
-    void processPsiFilesIfNecessary() {
+    public void processPsiFilesIfNecessary() {
+        if (psiFiles == null) {
+            return;
+        }
         if (!storage.isEnabled(activation)) {
-            LOGGER.info("Plugin not activated on " + project);
+            LOGGER.info("Action \"" + activation.getText() + "\" not enabled on " + project);
```
Comment (abimarank) [inline src/main/java/com/dubreuia/core/component/Engine.java:89]: Can we use `String.format()` here too as like in other places?

Comment (fishermans) [reply src/main/java/com/dubreuia/core/component/Engine.java]: @abimarank Thanks for reviewing. I replace all string concats in Engine class by String.format.

**Comment by dubreuia** [comment]
Thanks a lot @fishermans for those fixes 🙏 ! Merging and testing on my side, I'll try a release ASAP

Thank you to @abimarank for the review as well

**EDIT: tests OK 👍**

**Comment by dubreuia** [comment]
@fishermans - nitpicking on your pull request: you don't need to change the versions (like 2.3.1), I do it when I release I have a script which automates it (`./scripts/release.sh`)

Also I'm following `semver` so I don't increment the maintenance version for bugs / new features, here we go from 2.3.0 to 2.4.0. The version 2.3.1 would be a quick fix / hot fix or security patch (no new features)

**Comment by dubreuia** [comment]
https://github.com/dubreuia/intellij-plugin-save-actions/releases/tag/v2.4.0

**Comment by dubreuia** [comment]
Plugin uploaded pending verification, thanks again

**Comment by fishermans** [comment]
Thanks a lot @dubreuia! 

Maybe we can clean up some of the old tickets to have a better overview of what still needs to be done.

I guess, the option "Remove unused suppressed warning annotation" should be removed as long as this function is not working as expected. It removes too much at the moment and I haven't figured out how to fix it. See #87 .

**Pull request merged**